realtime.man
.TH "realtime" "1"
.SH NAME
realtime \- manage real-time configuration of systemd services
.SH SYNOPSIS
realtime [
.B -c
]
[
.B -l
]
[
.B -r
budget ]
[
.B -v
]
.SH OPTIONS
.TP
.B -c
Clean generated configuration, if any, and exit.
.TP
.B -l
Show default runtime budget and exit.
.TP
.BI -r\  runtime_us
Set default runtime budget to \fIruntime_us\fR and exit. This might not work when there already
exists a cgroup with a runtime budget in use on the system.
.TP
.B -v
Turn on verbose mode.
.SH DESCRIPTION
The utility generates drop-ins that allow services to assign real-time budget according to configuration.
.PP
First,
.B /etc/systemd/system/realtime-start.service
is created, which is necessary to initiate
.B realtime.slice
, which in turn is used as a root process group (cgroup) for services that need
real-time bugdet. All services managed by this utility are placed under it.
The
.B realtime-start.service
is started automatically and
.B realtime.slice
is assigned the default bugdet, called runtime_us, which is 950000 us.
.PP
Next, service drop-ins are generated according to configuration. Any drop-ins created during previous runs
are discarded.
.PP
.SH CONFIGURATION
Configuration is read from
.B /etc/systemd/realtime.conf
and from
.B /etc/systemd/realtime.conf.d/*.conf
and is expected to be in the following format:

    <service 1> : <runtime_us 1>
    <service 2> : <runtime_us 2>
    ...

Service names can be specified with or without the ".service" suffix.
Any empty lines are ignored. Items can be separated by any amount of white space.

Configuration files located in
.B /etc/systemd/realtime.conf.d/
are read in alphabetical order and must end with
\fB.conf\fR. If a service is assigned value multiple times, the last value updates the previous ones.

Note that the sum of all runtime_us values cannot be higher than runtime_us value set for realtime.slice.

.SH EXAMPLE
# cat /etc/systemd/realtime.conf

    cups  : 100000
    tftp  : 250000
    httpd : 400000

# cat /etc/systemd/realtime.conf.d/vsftpd.conf

    vsftpd : 200000

# realtime -v

    2018/08/13 10:43:24 creating /etc/systemd/system/vsftpd.service.d/99-realtime.conf for vsftpd.service
    2018/08/13 10:43:24 creating /etc/systemd/system/cups.service.d/99-realtime.conf for cups.service
    2018/08/13 10:43:24 creating /etc/systemd/system/tftp.service.d/99-realtime.conf for tftp.service
    2018/08/13 10:43:24 creating /etc/systemd/system/httpd.service.d/99-realtime.conf for httpd.service
    2018/08/13 10:43:24 reloading systemd


# systemctl start tftp

# systemctl status tftp

    ...
    Drop-In: /etc/systemd/system/tftp.service.d
             └─99-realtime.conf
    ...
    CGroup: /realtime.slice/tftp.service
            └─1879 /usr/sbin/in.tftpd -s /var/lib/tftpboot
.SH AUTHOR
Jan Synáček <jsynacek@redhat.com>

.SH SEE ALSO
https://www.kernel.org/doc/Documentation/scheduler/sched-rt-group.txt


realtime.go
package main

import (
	"bufio"
	"bytes"
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"os"
	"os/exec"
	"path"
	"path/filepath"
	"strconv"
)

const (
	DROPIN   = "99-realtime.conf"
	SLICE    = "realtime.slice"
	TEMPLATE = `# Generated realtime dropin
[Service]
Slice=realtime.slice
CPUAccounting=yes
ExecStartPre=/bin/bash -c 'echo %d > %s'
`
	START = `[Service]
Slice=realtime.slice
ExecStart=/bin/bash -c "echo >/dev/null"
CPUAccounting=yes
RemainAfterExit=yes
`
)

var (
	optVerbose    = flag.Bool("v", false, "verbose mode")
	optClean      = flag.Bool("c", false, "only clean generated drop-ins")
	optRuntime    = flag.Bool("l", false, "show default runtime budget")
	optSetRuntime = flag.Int("r", 0, "set default runtime budget")
)

type Config struct {
	data map[string]int
	sum  int
	max  int
}

func newConfig(maxRuntime int) *Config {
	return &Config{make(map[string]int), 0, maxRuntime}
}

func logInfo(format string, v ...interface{}) {
	if *optVerbose {
		log.Printf(format, v...)
	}
}

func reloadSystemd() {
	logInfo("reloading systemd")
	err := exec.Command("systemctl", "daemon-reload").Run()
	if err != nil {
		log.Fatalf("failed to reload systemd daemon: %v", err)
	}
}

// Create starting unit if it doesn't exist.
func createStartUnit() bool {
	file := "/etc/systemd/system/realtime-start.service"
	_, err := os.Stat(file)
	if err != nil { // Assume that the file doesn't exist.
		logInfo("creating realtime-start.service")
		err = ioutil.WriteFile(file, []byte(START), 0644)
		if err != nil {
			log.Fatalf("failed to write %s: %v", file, err)
		}
		return true
	}
	return false
}

func readDefault() string {
	b, err := ioutil.ReadFile("/sys/fs/cgroup/cpu/cpu.rt_runtime_us")
	if err != nil {
		log.Fatalf("failed to read: %v", err)
	}
	return string(bytes.TrimSpace(b))
}

func readDefaultInt() int {
	v, err := strconv.Atoi(readDefault())
	if err != nil {
		log.Fatalf("failed to convert runtime_us")
	}
	return v
}

func writeDefault(value int) {
	err := ioutil.WriteFile("/sys/fs/cgroup/cpu/cpu.rt_runtime_us", []byte(strconv.Itoa(value)), 0644)
	if err != nil {
		log.Fatalf("failed to write: %v", err)
	}
}

// Find unit in systemd directories.
func findUnit(name string) (unit string) {
	for _, dir := range []string{"/etc/systemd/system", "/usr/lib/systemd/system"} {
		unit = path.Join(dir, name)
		_, err := os.Stat(unit)
		if err == nil {
			return
		}
	}
	return ""
}

func unitName(unit string) (name string) {
	name = unit
	if path.Ext(name) != ".service" {
		name += ".service"
	}
	return
}

// Read configuration from a file.
// The format is "<service name> : <runtime budget>" per line. The values can be delimited
// by any amount of whitespace.
func (c *Config) readConfig(file string) error {
	f, err := os.Open(file)
	if err != nil {
		return err
	}
	defer f.Close()

	b := bufio.NewReader(f)
	l := 0
	for {
		line, err := b.ReadBytes('\n')
		if err == io.EOF {
			break
		}
		l++
		line = bytes.TrimSpace(line)
		if len(line) == 0 {
			continue
		}
		parts := bytes.Split(line, []byte(":"))
		if len(parts) != 2 {
			return fmt.Errorf("invalid configuration: %s:%d %q", file, l, string(line))
		}
		parts[0] = bytes.TrimSpace(parts[0])
		parts[1] = bytes.TrimSpace(parts[1])
		val, err := strconv.Atoi(string(parts[1]))
		if err != nil {
			return fmt.Errorf("invalid configuration: %s:%d %q", file, l, string(line))
		}

		name := unitName(string(parts[0]))
		unit := findUnit(name)
		if len(unit) == 0 {
			logInfo("%s does not exist, skipping", name)
			continue
		}

		v, ok := c.data[name]
		if ok {
			c.sum -= v
		}
		c.data[name] = val
		c.sum += val
	}
	return nil
}

func readConfig(maxRuntime int) (*Config, error) {
	cfg := newConfig(maxRuntime)
	err := cfg.readConfig("/etc/systemd/realtime.conf")
	if err != nil && !os.IsNotExist(err) {
		return nil, err
	}
	err = filepath.Walk("/etc/systemd/realtime.conf.d/", func(file string, info os.FileInfo, err error) error {
		if err != nil && !os.IsNotExist(err) {
			return err
		}
		if path.Ext(file) == ".conf" {
			err = cfg.readConfig(file)
			if err != nil {
				return err
			}
		}
		return nil
	})
	if cfg.sum == 0 {
		return nil, fmt.Errorf("refusing to generate empty (or zero budget) configuration")
	}
	if cfg.sum > cfg.max {
		return nil, fmt.Errorf("invalid configuration: runtime budget too high: %d > %d", cfg.sum, cfg.max)
	}
	return cfg, err
}

// Delete all realtime dropins possibly generated by this program.
func deleteDropins() error {
	return filepath.Walk("/etc/systemd/system/", func(file string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if path.Base(file) == DROPIN {
			logInfo("deleting %s", file)
			err = os.Remove(file)
			if err != nil {
				return err
			}
		}
		return nil
	})
}

// Apply the configuration.
func (c *Config) apply(rootRuntime int) {
	// Default runtime_us has to be copied down the hierarchy to the realtime slice.
	cgroup := path.Join("/sys/fs/cgroup/cpu", SLICE, "cpu.rt_runtime_us")
	err := ioutil.WriteFile(cgroup, []byte(strconv.Itoa(rootRuntime)), 0644)
	if err != nil {
		log.Fatalf("failed to write runtime_us: %v", err)
	}

	for service, runtime := range c.data {
		dropinDir := path.Join("/etc/systemd/system/", service+".d")
		err = os.Mkdir(dropinDir, 0644)
		if err != nil && !os.IsExist(err) {
			logInfo("failed to create directory: %v", err)
			continue
		}

		dropin := path.Join(dropinDir, "99-realtime.conf")
		cgroup = path.Join("/sys/fs/cgroup/cpu", SLICE, service, "cpu.rt_runtime_us")

		logInfo("creating %s for %s", dropin, service)
		err = ioutil.WriteFile(dropin, []byte(fmt.Sprintf(TEMPLATE, runtime, cgroup)), 0644)
		if err != nil {
			logInfo("failed to write to %s: %v", dropin, err)
		}
	}
}

func main() {
	flag.Parse()

	if *optClean {
		deleteDropins()
		return
	}
	if *optRuntime {
		fmt.Println(readDefault())
		return
	}
	if *optSetRuntime > 0 {
		writeDefault(*optSetRuntime)
		return
	}

	ok := createStartUnit()
	if ok {
		reloadSystemd()
	}

	err := exec.Command("systemctl", "start", "realtime-start").Run()
	if err != nil {
		log.Fatalf("failed to start realtime-start.service: %v", err)
	}

	rootRuntime := readDefaultInt()
	config, err := readConfig(rootRuntime)
	if err != nil {
		log.Fatal(err)
	}

	deleteDropins()
	config.apply(rootRuntime)
	reloadSystemd()
}
